File: minimum-spanning-trees/golang/prim.go
package main

import (
	"container/heap"
	"fmt"
	"math"
)

func prim(graph *Graph, r *Vertex) {
	vertexHeap := make(vertexHeap, 0)
	inHeap := make(map[string]bool)
	verticies := graph.vertices

	for _, vertex := range verticies {
		vertexHeap = append(vertexHeap, vertex)
		inHeap[vertex.name] = true
	}

	for _, v := range vertexHeap {
		if r.name != v.name {
			v.distance = math.MaxInt64
		} else {
			v.distance = 0
		}
	}

	heap.Init(&vertexHeap)

	adjacency_list := graph.adjacencyList
	for len(vertexHeap) > 0 {
		u := heap.Pop(&vertexHeap).(*Vertex)
		inHeap[u.name] = false

		adjacencies := adjacency_list[u.name]
		for _, adjacency := range adjacencies {
			v := verticies[adjacency.end]
			weight := adjacency.weight
			if inHeap[v.name] && v.distance > weight {
				fmt.Printf("u = %s, v = %s / before = %d, after = %d\n", u.name, v.name, v.distance, weight)

				v.distance = weight
				v.predecessor = u

				heap.Init(&vertexHeap)
			}
		}
	}
}

func main() {
	vertices := make(map[string]*Vertex)
	for _, name := range "abcdefghi" {
		vertices[string(name)] = &Vertex{name: string(name)}
	}

	edges := make([]*Edge, 0)
	edges = append(edges, &Edge{start: "a", end: "b", weight: 4})
	edges = append(edges, &Edge{start: "a", end: "h", weight: 8})
	edges = append(edges, &Edge{start: "b", end: "a", weight: 4})
	edges = append(edges, &Edge{start: "b", end: "c", weight: 8})
	edges = append(edges, &Edge{start: "b", end: "h", weight: 11})
	edges = append(edges, &Edge{start: "c", end: "b", weight: 8})
	edges = append(edges, &Edge{start: "c", end: "d", weight: 7})
	edges = append(edges, &Edge{start: "c", end: "f", weight: 4})
	edges = append(edges, &Edge{start: "c", end: "i", weight: 2})
	edges = append(edges, &Edge{start: "d", end: "c", weight: 7})
	edges = append(edges, &Edge{start: "d", end: "e", weight: 9})
	edges = append(edges, &Edge{start: "d", end: "f", weight: 14})
	edges = append(edges, &Edge{start: "e", end: "d", weight: 9})
	edges = append(edges, &Edge{start: "e", end: "f", weight: 10})
	edges = append(edges, &Edge{start: "f", end: "c", weight: 4})
	edges = append(edges, &Edge{start: "f", end: "d", weight: 14})
	edges = append(edges, &Edge{start: "f", end: "e", weight: 10})
	edges = append(edges, &Edge{start: "f", end: "g", weight: 2})
	edges = append(edges, &Edge{start: "g", end: "f", weight: 2})
	edges = append(edges, &Edge{start: "g", end: "h", weight: 1})
	edges = append(edges, &Edge{start: "g", end: "i", weight: 6})
	edges = append(edges, &Edge{start: "h", end: "a", weight: 8})
	edges = append(edges, &Edge{start: "h", end: "b", weight: 11})
	edges = append(edges, &Edge{start: "h", end: "g", weight: 1})
	edges = append(edges, &Edge{start: "h", end: "i", weight: 7})
	edges = append(edges, &Edge{start: "i", end: "c", weight: 2})
	edges = append(edges, &Edge{start: "i", end: "g", weight: 6})
	edges = append(edges, &Edge{start: "i", end: "h", weight: 7})

	graph := &Graph{vertices: vertices, edges: edges}
	graph.initializeGraph()

	fmt.Println("--- graph ---")
	graph.printGraph()

	fmt.Println("--- Prim's algorithm ---")
	prim(graph, vertices["a"])

	fmt.Println("--- result ---")
	total := 0
	for _, u := range vertices {
		fmt.Printf("u = %s", u.name)
		if u.predecessor != nil {
			fmt.Printf(", predecessor = %s", u.predecessor.name)
		}
		fmt.Printf(", weight = %d\n", u.distance)
		total += u.distance
	}
	fmt.Printf("total = %d\n", total)
}

File: minimum-spanning-trees/python/prim.py
import heapq
import sys

from graph import Edge, Graph, Vertex


def prim(graph, r):
    vertices = list(graph.vertices.values())
    adjacency_list = graph.adjacency_list

    for u in vertices:
        u.distance = sys.maxsize
        u.predecessor = None

    r.distance = 0

    heapq.heapify(vertices)

    while len(vertices):
        u = heapq.heappop(vertices)
        adjacencies = adjacency_list[u.name]
        for v, weight in adjacencies:
            if v in vertices and weight < v.distance:
                print(
                    f"u = {u.name}, v = {v.name} / before = {v.distance}, after = {weight}"
                )

                v.distance = weight
                v.predecessor = u

                heapq.heapify(vertices)


if __name__ == "__main__":
    vertices = {}
    for name in "abcdefghi":
        vertices[name] = Vertex(name)

    edges = []
    edges.append(Edge("a", "b", 4))
    edges.append(Edge("a", "h", 8))
    edges.append(Edge("b", "a", 4))
    edges.append(Edge("b", "c", 8))
    edges.append(Edge("b", "h", 11))
    edges.append(Edge("c", "b", 8))
    edges.append(Edge("c", "d", 7))
    edges.append(Edge("c", "f", 4))
    edges.append(Edge("c", "i", 2))
    edges.append(Edge("d", "c", 7))
    edges.append(Edge("d", "e", 9))
    edges.append(Edge("d", "f", 14))
    edges.append(Edge("e", "d", 9))
    edges.append(Edge("e", "f", 10))
    edges.append(Edge("f", "c", 4))
    edges.append(Edge("f", "d", 14))
    edges.append(Edge("f", "e", 10))
    edges.append(Edge("f", "g", 2))
    edges.append(Edge("g", "f", 2))
    edges.append(Edge("g", "h", 1))
    edges.append(Edge("g", "i", 6))
    edges.append(Edge("h", "a", 8))
    edges.append(Edge("h", "b", 11))
    edges.append(Edge("h", "g", 1))
    edges.append(Edge("h", "i", 7))
    edges.append(Edge("i", "c", 2))
    edges.append(Edge("i", "g", 6))
    edges.append(Edge("i", "h", 7))

    graph = Graph(vertices, edges)

    print("--- graph ---")
    graph.print_graph()

    print("--- Prim's algorithm ---")
    prim(graph, vertices["a"])

    print("--- result ---")
    for u in vertices.values():
        print(
            f"vertices = {u.name}, predecessor = {'None' if not u.predecessor else u.predecessor.name}, weight = {u.distance}"
        )
